Feature #21-QuingService

chanwooo opened the pull request:
QuingService 일부 구현중

Todo
나머지 구현
테스트  작성

chanwooo committed:
#21-변수명 변경 phone -> phoneNumber

chanwooo committed:
#21-타입변경 Integer -> int

chanwooo committed:
#21-안쓰는 파일 삭제

chanwooo committed:
#21-사용하지않는 필드 제거

chanwooo committed:
#21-QuingService 일부 구현
append
getList
countForward
doneWaiting

chanwooo commented:
서비스가 여러개 얽히니까 unit test보다 integration test가 동작을 확인하기에 편한거같네요 이런경우에는 둘 다 작성해야겠죠?

f-lab-trey commented on src/main/java/flab/quing/waiting/QuingServiceImpl.java:
```diff
@@ -0,0 +1,75 @@
+
+    @Override
+    public List<WaitingResponse> getList(long storeId) {
+//        return waitingRepository.findAllByStoreId(storeId)
```
지워도 될 거 같습니다.

f-lab-trey commented on src/main/java/flab/quing/waiting/QuingServiceImpl.java:
```diff
@@ -0,0 +1,75 @@
+
+    @Override
+    @Transactional
+    public WaitingResponse append(WaitingRequest waitingRequest) {
```
[q]
한 사용자가 중복으로 줄서기 하는 경우는 어떻게 처리되나요? 이 메서드에는 방어 로직이 없는 거 같은데..

f-lab-trey commented on src/main/java/flab/quing/waiting/QuingServiceImpl.java:
```diff
@@ -0,0 +1,75 @@
+    public int countForward(long waitingId) {
+        Waiting waiting = waitingRepository.findById(waitingId).orElseThrow(NoSuchWaitingException::new);
+        List<WaitingResponse> list = getList(waiting.getStore().getId());
+        return list.indexOf(waiting.toResponse());
```
이 operation의 시간 복잡도는 어떻게 될까요?

f-lab-trey commented:
> 서비스가 여러개 얽히니까 unit test보다 integration test가 동작을 확인하기에 편한거같네요 이런경우에는 둘 다 작성해야겠죠?

네 둘 다 작성하는게 좋고, 둘 중 하나만 가능하다면 저는 UT를 추가하는 것이 더 좋을 거 같습니다. IT는 실행하는 데 시간이 오래걸리고, dependency로 인해 실패하는 경우가 많아서 나중에 관리하기가 힘든 단점이 있습니다.

chanwooo committed:
#21-QuingService test 일부 구현

chanwooo committed:
#21-주석 삭제

chanwooo committed:
#21-append 중복 방지 로직 추가

chanwooo committed:
#21-waiting cancel 구현

chanwooo committed:
#21-exception 패키지 이동

chanwooo committed:
#21-test코드 작성

chanwooo commented:
- append 중복 방지 로직 추가
- waiting cancel, done 구현
- test 작성

@f-lab-trey 리뷰부탁드릴게요

f-lab-trey commented on src/main/java/flab/quing/waiting/QuingServiceImpl.java:
```diff
@@ -0,0 +1,83 @@
+        Store store = storeRepository.findById(waitingRequest.getStoreId()).orElseThrow(NoSuchStoreException::new);
+        Waiting waiting = Waiting.of(user, store);
+
+        Optional<Waiting> exist = waitingRepository.findByUserIdAndWaitingQueueStatusIs(user.getId(), WaitingQueueStatus.WAITING);
```
[q]
이 operation의 시간복잡도는 어떻게 될까요? 중복을 체크하기 위해 너무 비싼 operation을 하는 게 아닐까요?

f-lab-trey commented on src/test/java/flab/quing/waiting/QuingServiceImplTest.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    @Test
+    void append_exist_throw_exception() {
```
[minor]
테스트 이름을 camel case로 바꿔주세요.

f-lab-trey commented on src/test/java/flab/quing/waiting/QuingServiceImplTest.java:
```diff
@@ -0,0 +1,141 @@
+        when(waitingRepository.save(any())).thenReturn(waiting);
+
+        WaitingResponse waitingResponse = quingService.append(waitingRequest);
+        System.out.println("waitingResponse = " + waitingResponse);
```
logger로 변경해주세요.

chanwooo committed:
Feature #22-waiting - 사용자에게 메시지 send 구현 (#27)

* #21-cancel 리턴 타입 변경

* #22-message 전송 관련 부분 print로 임시 구현

* #22-message 전송 관련 부분 test 작성

* #22-test에 불필요한 공백 제거

* #22- test코드 logger 적용

* #22- test코드 given, when, then 분리

* #22- test코드 logger 적용

* #22-리뷰반영
- 테스트 공백, 주석 정리
- 테스트 이름 변경 appendExistThrowDuplicateException

chanwooo committed:
# Conflicts:
#	src/main/java/flab/quing/user/User.java
#	src/main/java/flab/quing/user/UserRepository.java

chanwooo committed:
Merge remote-tracking branch 'origin/feature-#21' into feature-#21

chanwooo committed:
Merge branch 'develop' into feature-#21

# Conflicts:
#	src/main/java/flab/quing/user/User.java
#	src/main/java/flab/quing/user/UserRepository.java

chanwooo committed:
#21-user response import
